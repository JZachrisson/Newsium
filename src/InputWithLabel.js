import React from 'react';

const InputWithLabel = ({
  id,
  value,
  onInputChange,
  type = 'text',
  isFocused,
  children,
}) => {
  const inputRef = React.useRef();

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <div>
      <label htmlFor={id} className="label">
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
        className="input"
      />
    </div>
  );
};

export default InputWithLabel;
