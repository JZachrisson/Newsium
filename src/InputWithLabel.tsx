import React from 'react';
import { StyledLabel, StyledInput } from './styles';

type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  isFocused?: boolean;
  children: React.ReactNode;
};

const InputWithLabel = ({
  id,
  value,
  onInputChange,
  type = 'text',
  isFocused,
  children,
}: InputWithLabelProps) => {
  const inputRef = React.useRef<HTMLInputElement>(null!);

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <div>
      <StyledLabel htmlFor={id}>{children}</StyledLabel>
      &nbsp;
      <StyledInput
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </div>
  );
};

export default InputWithLabel;
