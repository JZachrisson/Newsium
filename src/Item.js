import React from 'react';

const Item = ({ item, onRemoveItem }) => {
  const handleRemoveItem = () => {
    onRemoveItem(item);
  };

  return (
    <div>
      <span>
        <a href={item.url}>{item.title}</a>
      </span>
      <span> {item.author}</span>
      <span> {item.num_comments}</span>
      <span> {item.points}</span>
      <span>
        <button type="button" onClick={handleRemoveItem}>
          Dismiss
        </button>
      </span>
    </div>
  );
};

export default Item;
