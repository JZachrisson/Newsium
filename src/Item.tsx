import React from 'react';
import styles from './App.module.css';
import cs from 'classnames';
import { StyledItem, StyledButtonSmall } from './styles';
import { ReactComponent as Check } from './check.svg';

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => {
  const handleRemoveItem = () => {
    onRemoveItem(item);
  };

  return (
    <StyledItem>
      <span style={{ width: '40%' }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: '30%' }}> {item.author}</span>
      <span style={{ width: '10%' }}> {item.num_comments}</span>
      <span style={{ width: '10%' }}> {item.points}</span>
      <span style={{ width: '10%' }}>
        <StyledButtonSmall
          type="button"
          onClick={handleRemoveItem}
          className={cs(styles.button, styles.buttonSmall)}
        >
          Dismiss
          <Check height="18px" width="18px" />
        </StyledButtonSmall>
      </span>
    </StyledItem>
  );
};

export default Item;
