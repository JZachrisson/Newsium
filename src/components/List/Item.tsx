import React from 'react';
import moment from 'moment';
import { StyledItem, StyledButtonSmall, StyledColumn } from '../../styles';
import { ReactComponent as Check } from '../../assets/check.svg';

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => {
  const handleRemoveItem = () => {
    onRemoveItem(item);
  };

  return (
    <StyledItem>
      <StyledColumn className="show" width="40%">
        <a href={item.url}>{item.title}</a>
      </StyledColumn>

      <StyledColumn width="20%">
        {' '}
        {moment.unix(item.created_at_i).format('DD-MM-YYYY')}
      </StyledColumn>
      <StyledColumn width="10%"> {item.author}</StyledColumn>
      <StyledColumn className="hide" width="10%">
        {' '}
        {item.num_comments}
      </StyledColumn>
      <StyledColumn className="hide" width="10%">
        {' '}
        {item.points}
      </StyledColumn>
      <StyledColumn className="hide" width="10%">
        <StyledButtonSmall type="button" onClick={handleRemoveItem}>
          Dismiss
          <Check height="18px" width="18px" />
        </StyledButtonSmall>
      </StyledColumn>
    </StyledItem>
  );
};

export default Item;
