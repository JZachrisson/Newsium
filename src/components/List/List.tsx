import React from 'react';
import Item from './Item';
import { StyledButtonSmall } from '../../styles';
import { sortBy } from 'lodash';

const SORTS: { [list: string]: any } = {
  NONE: (list: any) => list,
  TITLE: (list: any) => sortBy(list, 'title'),
  AUTHOR: (list: any) => sortBy(list, 'author'),
  DATE: (list: any) => sortBy(list, 'created_at'),
  COMMENT: (list: any) => sortBy(list, 'num_comments').reverse(),
  POINT: (list: any) => sortBy(list, 'points').reverse(),
};

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};

const List = ({ list, onRemoveItem }: ListProps) => {
  const [sort, setSort] = React.useState({
    sortKey: 'NONE',
    isReverse: false,
  });

  const handleSort = (sortKey: any) => {
    const isReverse = sort.sortKey === sortKey && !sort.isReverse;

    setSort({ sortKey, isReverse });
  };

  const sortFunction = SORTS[sort.sortKey];
  const sortedList = sort.isReverse
    ? sortFunction(list).reverse()
    : sortFunction(list);

  return (
    <div>
      <div style={{ display: 'flex' }}>
        <span style={{ width: '40%' }}>
          <StyledButtonSmall
            style={{ marginTop: '10px' }}
            type="button"
            onClick={() => handleSort('TITLE')}
          >
            Title
          </StyledButtonSmall>
        </span>
        <span style={{ width: '20%' }}>
          <StyledButtonSmall type="button" onClick={() => handleSort('DATE')}>
            Date
          </StyledButtonSmall>
        </span>
        <span style={{ width: '10%' }}>
          {' '}
          <StyledButtonSmall type="button" onClick={() => handleSort('AUTHOR')}>
            Author
          </StyledButtonSmall>
        </span>
        <span className="hide" style={{ width: '10%' }}>
          {' '}
          <StyledButtonSmall
            type="button"
            onClick={() => handleSort('COMMENT')}
          >
            Comments
          </StyledButtonSmall>
        </span>
        <span className="hide" style={{ width: '10%' }}>
          <StyledButtonSmall type="button" onClick={() => handleSort('POINT')}>
            Points
          </StyledButtonSmall>
        </span>
        <span className="hide" style={{ width: '10%' }}>
          Actions
        </span>
      </div>
      {sortedList.map((item: any) => (
        <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
      ))}
    </div>
  );
};

export default List;
