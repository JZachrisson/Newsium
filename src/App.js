import React from 'react';
import List from './List';
import useSemiPersistentState from './useSemiPersistentState';
import storiesReducer from './storiesReducer';
import SearchForm from './SearchForm';
import axios from 'axios';
import styles from './App.module.css';

const API_ENDPOINT = 'http://hn.algolia.com/api/v1/search?query=';

function App() {
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const handleFetchStories = React.useCallback(async () => {
    if (searchTerm === '') return;

    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    try {
      const result = await axios.get(url);

      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [url]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  return (
    <div className={styles.container}>
      <h1 className={styles.headlinePrimary}>My Hacker Stories</h1>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
}

export default App;
