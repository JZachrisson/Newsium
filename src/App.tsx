import React from 'react';
import List from './List';
import useSemiPersistentState from './useSemiPersistentState';
import storiesReducer from './storiesReducer';
import SearchForm from './SearchForm';
import axios from 'axios';
import { StyledContainer, StyledHeadlinePrimary } from './styles';

const API_ENDPOINT = 'http://hn.algolia.com/api/v1/search?query=';

const getUrl = (searchTerm: string) => `${API_ENDPOINT}${searchTerm}`;

const extractSearchTerm = (url: string) => url.replace(API_ENDPOINT, '');

const getLastSearches = (urls: string[]) =>
  urls.slice(-5).map(extractSearchTerm);

function App() {
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });
  const [urls, setUrls] = React.useState([getUrl(searchTerm)]);

  console.log('B:A');

  const handleFetchStories = React.useCallback(async () => {
    if (searchTerm === '') return;

    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    try {
      const lastUrl = urls[urls.length - 1];
      const result = await axios.get(lastUrl);

      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [urls]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = (item: Story) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    handleSearch(searchTerm);
    event.preventDefault();
  };

  const handleLastSearch = (searchTerm: string) => {
    handleSearch(searchTerm);
  };

  const handleSearch = (searchTerm: string) => {
    const url = getUrl(searchTerm);
    setUrls(urls.concat(url));
  };

  const lastSearches = getLastSearches(urls);

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories.</StyledHeadlinePrimary>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {lastSearches.map((url) => (
        <button
          key={searchTerm}
          type="button"
          onClick={() => handleLastSearch(searchTerm)}
        >
          {searchTerm}
        </button>
      ))}

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </StyledContainer>
  );
}

export default App;
