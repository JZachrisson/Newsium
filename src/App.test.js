import React from 'react';
import renderer from 'react-test-renderer';

import App from './App';
import Item from './Item';
import List from './List';
import SearchForm from './SearchForm';
import InputWithLabel from './InputWithLabel';

describe('Item', () => {
  const item = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };

  it('renders all properties', () => {
    const component = renderer.create(<Item item={item} />);

    expect(component.root.findByType('a').props.href).toEqual(
      'https://reactjs.org/'
    );

    expect(component.root.findAllByProps({ children: 'React' }).length).toEqual(
      1
    );
  });
});
