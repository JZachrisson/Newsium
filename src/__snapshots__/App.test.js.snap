// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  className="sc-dlfnbm gebkuo"
>
  <span
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style={
      Object {
        "width": "30%",
      }
    }
  >
     
    Jordan Walke
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
     
    3
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
     
    4
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="sc-hKgILt sc-eCssSg bmEmLV kjabvn button buttonSmall"
      onClick={[Function]}
      type="button"
    >
      Dismiss
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</div>
`;
